src/org/kms/core/KMSUtils.java

```diff
@@ -0,0 +1,32 @@
+package org.kms.core;
+
+import java.util.HashSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class KMSUtils {
+
+	private static HashSet<String>lockSet = new HashSet<String>();
+	private static Lock lock = new ReentrantLock();
+
+	static HashSet<String> getMap(){
+		return lockSet;
+	}
+
+	public static boolean putWithLock(String id){
+		if (lockSet.contains(id)){
+			return false;
```

Comment:
what is the purpose of these two lines ? this condition is checked in the else part by taking a lock first